### PR TITLE
Fully use circe to parse TensorizeIn config

### PR DESCRIPTION
**Pull request by sacsar**
**Background**: #31 moved to parsing the TensorizeIn config as a Typesafe Config object. However, this meant that the `InputFeatureInfo` had to be decoded manually because the transform config had type `Map[String,Map[String,Any]]`. (See [this discussion](https://github.com/linkedin/Avro2TF/pull/31#discussion_r296965348))

**This pull request does the following**:
- introduces `TransformConfig`, `HashInfo` and `Tokenization` case classes
- removes the special handling for `InputFeatureInfo` when the config is parsed
- changes the type of the cardinality field in `TensorMetadata` from `Option[Any]` to `Option[Long]`
- parses the `dtype` and `combiner` fields directly into the enums rather than passing through the string

**Why make these changes?**
Replacing the use of `Map[String,Any]` with the transform config case classes opens the possibility for validating the config json (currently, if the key to a field is misspelled, the field is dropped, which is behavior retained in this change). It also means that the behavior of the configuration is fully reflected in the code.

### CODE REVIEW

File: avro2tf/src/main/scala/com/linkedin/avro2tf/configs/TensorizeInConfig.scala [hunk 1]
```diff
@@ -23,82 +23,97 @@ object CombinerType extends Enumeration {
...
- * @param combinerType Which combiner to use when collision happens
+ * @param numHashFunctions The number of hash functions to use - defaults to 1
+ * @param combiner Which combiner to use when collision happens
  */
 case class HashInfo(
-  hashBucketSize: Int,
-  numHashFunctions: Int = 1,
-  combinerType: CombinerType.CombinerType = CombinerType.SUM
-)
+                     hashBucketSize: Int,
```
Comment (zhangxuhong) [inline avro2tf/src/main/scala/com/linkedin/avro2tf/configs/TensorizeInConfig.scala]: can you fixed the style here? we use photon-ml style, you can import [this](https://github.com/linkedin/photon-ml/blob/master/Photon%20Style.xml) code style for scala into intellij

File: avro2tf/src/main/scala/com/linkedin/avro2tf/configs/TensorizeInConfig.scala [hunk 2]
```diff
@@ -23,82 +23,97 @@ object CombinerType extends Enumeration {
...
+  * Case class for TensorizeIn configuration
+  *
+  * @param features A sequence of features
+  * @param labels   A sequence of labels, which may be empty
+  */
 case class TensorizeInConfiguration(
-  features: Seq[Feature],
-  labels: Option[Seq[Feature]]
-) {
+                                     features: Seq[Feature],
```
Comment (zhangxuhong) [inline avro2tf/src/main/scala/com/linkedin/avro2tf/configs/TensorizeInConfig.scala]: fix style

File: avro2tf/src/main/scala/com/linkedin/avro2tf/configs/TensorizeInConfig.scala (hunk 1 above)
Comment (sacsar) [reply avro2tf/src/main/scala/com/linkedin/avro2tf/configs/TensorizeInConfig.scala]: Thanks. Let me whack everything with the formatter quickly.

File: avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/TensorMetadataGeneration.scala [hunk 3]
```diff
@@ -93,7 +97,13 @@ class TensorMetadataGeneration {
    */
   private def getColsWithHashInfoCardinalityMapping(params: TensorizeInParams): Map[String, Long] = {
 
-    TensorizeInConfigHelper.getColsHashInfo(params).map(x => x._1 -> x._2.hashBucketSize.toLong)
+    TensorizeInConfigHelper.concatFeaturesAndLabels(params)
```
Comment (zhangxuhong) [inline avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/TensorMetadataGeneration.scala]: why not reuse `TensorizeInConfigHelper.getColsHashInfo`?

**Review by zhangxuhong — CHANGES_REQUESTED** [review]
Overall, LGTM. Thanks for taking time to improve the code quality of Avro2TF. Just one minor thing on the code style, please follow photon-ml style as i explained in the comments.

File: avro2tf/src/main/scala/com/linkedin/avro2tf/configs/Codecs.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package com.linkedin.avro2tf.configs
+
+import io.circe.{Decoder, Encoder}
+import io.circe.generic.extras.Configuration
+
+/**
+  * Implicits for Circe.
```
Comment (cyzhangchenya) [inline avro2tf/src/main/scala/com/linkedin/avro2tf/configs/Codecs.scala]: Do you think we can rename this file `Codecs.scala` to a more meaningful one?

File: avro2tf/src/main/scala/com/linkedin/avro2tf/helpers/HashingTransformer.scala [hunk 5]
```diff
@@ -81,14 +81,14 @@ object HashingTransformer {
               group => {
                 val idValue = group._2.reduce {
                   (a, b) => {
-                    hashInfo.combinerType match {
+                    hashInfo.combiner match {
                       case CombinerType.SUM | CombinerType.AVG => a.copy(value = a.value + b.value)
```
Comment (cyzhangchenya) [inline avro2tf/src/main/scala/com/linkedin/avro2tf/helpers/HashingTransformer.scala]: Do we also want to update the enum name of `CombinerType` to `Combiner`?

File: avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/TensorMetadataGeneration.scala (hunk 3 above)
Comment (sacsar) [reply avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/TensorMetadataGeneration.scala]: That's definitely better. Not sure what I was thinking.

**Comment by sacsar** [comment]
@zhangxuhong There are conflicts with master at the moment. Would you prefer I rebase onto master, merge master or have the person who commits resolve the conflicts?

File: avro2tf/src/test/scala/com/linkedin/avro2tf/parsers/TensorizeInConfigParserTest.scala [hunk 6]
```diff
@@ -1,18 +1,21 @@
 package com.linkedin.avro2tf.parsers
 
+import com.linkedin.avro2tf.configs.{CombinerType, HashInfo, Tokenization}
 import com.typesafe.config.ConfigException
 
 import scala.io.Source
 import org.testng.Assert._
 import org.testng.annotations.Test
-
+import io.circe.generic.extras.auto._
```
Comment (cyzhangchenya) [inline avro2tf/src/test/scala/com/linkedin/avro2tf/parsers/TensorizeInConfigParserTest.scala]: Please reorder the import.

**Comment by zhangxuhong** [comment]
@sacsar Please rebase to master

**Review by cyzhangchenya — COMMENTED** [review]
Thanks Sebastian for the contribution! :)

File: avro2tf/src/main/scala/com/linkedin/avro2tf/configs/Codecs.scala (hunk 4 above)
Comment (sacsar) [reply avro2tf/src/main/scala/com/linkedin/avro2tf/configs/Codecs.scala]: I'm renaming it to `JsonCodecs`, which is a bit better.

File: avro2tf/src/main/scala/com/linkedin/avro2tf/helpers/HashingTransformer.scala (hunk 5 above)
Comment (sacsar) [reply avro2tf/src/main/scala/com/linkedin/avro2tf/helpers/HashingTransformer.scala]: I think so. There's a similar inconsistency with `dtype` vs `datatype`, but that inconsistency exists in Tensorflow as well.

File: avro2tf/src/test/scala/com/linkedin/avro2tf/parsers/TensorizeInConfigParserTest.scala (hunk 6 above)
Comment (sacsar) [reply avro2tf/src/test/scala/com/linkedin/avro2tf/parsers/TensorizeInConfigParserTest.scala]: Done.

**Review by cyzhangchenya — APPROVED** [review]
Thanks Sebastian.

**Review by zhangxuhong — APPROVED** [review]

**Pull request merged**